test/opt/test_def_use.cpp

```diff
@@ -0,0 +1,599 @@
+  CheckDefUse(tc.du, manager.id_to_defs(), manager.id_to_uses());
+}
+
+// clang-format off
+INSTANTIATE_TEST_CASE_P(
+    TestCase, ReplaceUseTest,
+    ::testing::ValuesIn(std::vector<ReplaceUseCase>{
+      { // no use, no replace request
+        "", {}, "", {},
+      },
+      { // no use, some replace requests
+        "OpMemoryModel Logical GLSL450",
+        {{1, 2}, {3, 4}, {7, 8}, {7, 9}, {7, 10}, {2, 10}, {3, 10}},
+        "OpMemoryModel Logical GLSL450",
+        {},
+      },
+      { // replace one use
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 3",
+        {{1, 3}},
+        "%1 = OpTypeBool\n"
+        "%2 = OpTypeVector %3 3",
+        {
+          {
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %3 3"},
+          },
+          {
+            {3, {"%2 = OpTypeVector %3 3"}},
+          },
+        },
+      },
+      { // replace and then replace back
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 3",
+        {{1, 3}, {3, 1}},
+        "%1 = OpTypeBool\n"
+        "%2 = OpTypeVector %1 3",
+        {
+          {
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %1 3"},
+          },
+          {
+            {1, {"%2 = OpTypeVector %1 3"}},
+          },
+        },
+      },
+      { // replace in sequence
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 3",
+        {{1, 3}, {3, 4}, {4, 5}, {5, 100}},
+        "%1 = OpTypeBool\n"
+        "%2 = OpTypeVector %100 3",
+        {
+          {
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %100 3"},
+          },
+          {
+            {100, {"%2 = OpTypeVector %100 3"}},
+          },
+        },
+      },
+      { // replace multiple uses
+        "%1 = OpTypeBool "
+        "%2 = OpTypeVector %1 2 "
+        "%3 = OpTypeVector %1 3 "
+        "%4 = OpTypeVector %1 4 "
+        "%5 = OpTypeMatrix %2 2 "
+        "%6 = OpTypeMatrix %3 3 "
+        "%7 = OpTypeMatrix %4 4",
+        {{1, 10}, {2, 20}, {4, 40}},
+        "%1 = OpTypeBool\n"
+        "%2 = OpTypeVector %10 2\n"
+        "%3 = OpTypeVector %10 3\n"
+        "%4 = OpTypeVector %10 4\n"
+        "%5 = OpTypeMatrix %20 2\n"
+        "%6 = OpTypeMatrix %3 3\n"
+        "%7 = OpTypeMatrix %40 4",
+        {
+          {
+            {1, "%1 = OpTypeBool"},
+            {2, "%2 = OpTypeVector %10 2"},
+            {3, "%3 = OpTypeVector %10 3"},
+            {4, "%4 = OpTypeVector %10 4"},
+            {5, "%5 = OpTypeMatrix %20 2"},
+            {6, "%6 = OpTypeMatrix %3 3"},
+            {7, "%7 = OpTypeMatrix %40 4"},
+          },
+          {
+            {10,
+              {
+                "%2 = OpTypeVector %10 2",
+                "%3 = OpTypeVector %10 3",
+                "%4 = OpTypeVector %10 4",
+              }
+            },
+            {20, {"%5 = OpTypeMatrix %20 2"}},
+            {3, {"%6 = OpTypeMatrix %3 3"}},
+            {40, {"%7 = OpTypeMatrix %40 4"}},
+          },
```

Comment:
Another test case:
  Replace a case Id in OpSwitch over 64-bit integers.  (other than the default case)
This is interesting because it will check that you are replacing the right word when the literals are 2 words wide.

(We also don't have lower level tests for SetOperand and GetSingleWordOperand.  But at least a case here will check it at slightly above unit test level.)